tests/test_sampler.py

```diff
@@ -34,121 +33,93 @@ def test_sample_rate_deviation(self):
-
-            assert got == expected, \
-                "Wrong number of traces sampled, %s instead of %s" % (got, expected)
-
-    def test_long_run(self):
-        writer = DummyWriter()
         tracer = Tracer()
         tracer.writer = writer
 
-        # Test a big matrix of combinaisons
-        # Ensure to have total_time >> BUFFER_DURATION to reduce edge effects
-        for tps in [10, 23, 15, 31]:
-            for (traces_per_s, total_time) in [(80, 23), (75, 66), (1000, 77)]:
-
-                with patch_time() as fake_time:
-                    # We do tons of operations in this test, do not let the time slowly shift
-                    fake_time.set_delta(0)
-
-                    tracer.sampler = ThroughputSampler(tps)
-
-                    for _ in range(total_time):
-                        for _ in range(traces_per_s):
-                            s = tracer.trace("whatever")
-                            s.finish()
-                        fake_time.sleep(1)
+        tracer.sampler = RateSampler(0.5)
 
-                traces = writer.pop()
-                # The current sampler implementation can introduce an error of up to
-                # `tps * BUFFER_DURATION` traces at initialization (since the sampler starts empty)
-                got = len(traces)
-                expected = tps * total_time
-                error_delta = tps * tracer.sampler.BUFFER_DURATION
+        random.seed(1234)
 
-                assert abs(got - expected) <= error_delta, \
-                    "Wrong number of traces sampled, %s instead of %s (error_delta > %s)" % (got, expected, error_delta)
+        for i in range(10):
+            span = tracer.trace(i)
+            span.finish()
 
+            samples = writer.pop()
+            assert len(samples) <= 1, "there should be 0 or 1 spans"
+            sampled = (1 == len(samples))
+            for j in range(10):
+                other_span = Span(tracer, i, trace_id=span.trace_id)
+                assert sampled == tracer.sampler.sample(other_span), "sampling should give the same result for a given trace_id"
 
-    def test_concurrency(self):
-        # Test that the sampler works well when used in different threads
+class RateByServiceSamplerTest(unittest.TestCase):
+    def test_sample_rate_deviation(self):
         writer = DummyWriter()
-        tracer = Tracer()
-        tracer.writer = writer
 
-        total_time = 3
-        concurrency = 100
-        end_time = time.time() + total_time
+        for sample_rate in [0.1, 0.25, 0.5, 1]:
+            tracer = Tracer()
+            tracer.configure(sampler=AllSampler(), priority_sampling=True)
+            tracer.priority_sampler.set_sample_rate(sample_rate)
+            tracer.writer = writer
 
-        # Let's sample to a multiple of BUFFER_SIZE, so that we can pre-populate the buffer
-        tps = 15 * ThroughputSampler.BUFFER_SIZE
-        tracer.sampler = ThroughputSampler(tps)
+            random.seed(1234)
```

Comment:
To reduce flakyness, indeed. A througput sampler is based on trace IDs which are pseudo-random so indeed with bad luck even if tuned at say 3% it could behave like it's 30%. This should be rare but again, that's the definition of flakiness.